### PR TITLE
[engine] Adapt soft-upgrade integration-tests

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances: #17082

Adapt soft-upgrade-integration-tests to match latest required behaviour.
- We expect success for any types which satisfy the compatibility relation. And specifically, in the case of downgrades.

Fix existing implementation.
- This implementation will be completely reworked soon. The _drop/make-None_ needs to be done during `importValue`. 
- Take this opportunity to remove `SBPromoteAnyContract`, and collapse the behaviour into `SBCastAnyContract`. This is in preparation for control by an engine feature-flag.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 1]
```diff
@@ -187,8 +187,13 @@ object SValue {
 
   object SRecord {
     def apply(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue]): SRecord = {
-      assert(fields.length == values.size)
-      val zipped = (fields.toSeq zip values.asScala).toList.reverse
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala:191]: This `reverse` was left in by accident on the previous PR. It was just a development-time test to ensure nothing depended on the order any longer.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1075,88 +1074,63 @@ private[lf] object SBuiltin {
...
-          case (name @ _, TOptional(_)) => false
-          case _ => true
-        }
-        if (badNewFields.isEmpty) {
-          Control.Value(
+      if (actualTemplateId == templateId) {
+        // No upgrade or downgrade required
+        Control.Value(record)
+      } else {
+        assert(actualTemplateId != templateId)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1098]: Given we are within a conditional, do we need this assertion?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -1075,88 +1074,63 @@ private[lf] object SBuiltin {
...
+        if (!allowUpgradesAndDowngrades) {
+          // Existing behavior retained when the up/down-grades feature is disabled.
+          val coid = getSContractId(args, 0)
+          Control.Error(IE.WronglyTypedContract(coid, templateId, actualTemplateId))
+        } else {
+          assert(allowUpgradesAndDowngrades)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1104]: Likewise, do we need the assertion?

**Review by carlpulley-da — APPROVED** [review]
Other than some minor comments that might be debugging cruft(?), LGTM!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1098]: No, we don't need these. 
I was just being careful and trying to make the code easier to follow.

**Comment by nickchapman-da** [comment]
Thanks for your review!

**Pull request merged**